tests/unit/drivers/test_reduce_all_driver.py

```diff
@@ -0,0 +1,63 @@
+import os
+import pytest
+from typing import List, Dict
+import numpy as np
+from jina.flow import Flow
+from jina.proto.jina_pb2 import Document
+from jina.executors.crafters import BaseSegmenter
+from jina.executors.encoders import BaseEncoder
+from tests import JinaTestCase, random_docs
+
+cur_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+class MockSegmenter(BaseSegmenter):
+
+    def craft(self, text: str, *args, **kwargs) -> List[Dict]:
+        split = text.split(',')
+        chunks = [dict(text=split[0], offset=0, weight=1.0, mode_id='mode1'),
+                  dict(text=split[1], offset=1, weight=1.0, mode_id='mode2')]
+        return chunks
+
+
+class MockEncoder(BaseEncoder):
+
+    def encode(self, data: str, *args, **kwargs) -> 'np.ndarray':
+        output = []
+        for r in data:
+            if "mode1" in r:
+                output.append([0.0, 0.0, 0.0])
+            elif "mode2" in r:
+                output.append([1.0, 1.0, 1.0])
+
+        return np.array(output)
+
+
+class ReduceAllDriverTestCase(JinaTestCase):
+    @pytest.mark.skip('Not working as expected')
+    def test_merge_chunks_with_different_mode_id(self):
```

Comment:
@hanxiao, I was thinking about a test like this one. after reducing, I would pretend all the documents have all their chunks.

Reply:
ok let me check

Reply:
what does this do? this part feels strange, given your doc text, it basically encodes everything into `[0.0, 0.0, 0.0]`
```
    def encode(self, data: str, *args, **kwargs) -> 'np.ndarray':
        output = []
        for r in data:
            if "mode1" in r:
                output.append([0.0, 0.0, 0.0])
            elif "mode2" in r:
                output.append([1.0, 1.0, 1.0])

        return np.array(output)
```

Reply:
ok first observation is that doc_pointer mixes docs from multiple `depth_level` together, which is always dangerous. Let me check how this affects the result

![image](https://user-images.githubusercontent.com/2041322/88770883-bb7f7b80-d17e-11ea-80cd-6d1a4672a9c4.png)

Reply:
and in `ReduceAllDriver`, this "force a shallow copy" does not guarantee it copies from the docs in the current request
```
class ReduceAllDriver(ReduceDriver):
    """:class:`ReduceAllDriver` merges chunks and matches from all requests """

    def _apply(self, doc: 'jina_pb2.Document', *args, **kwargs):
        if doc.id not in self.doc_pointers:
            self.doc_pointers[doc.id] = copy(doc)  # force a shallow copy
```

Reply:
ok let me fix `ReduceAllDriver` on this branch

Reply:
is it now working this test?

Reply:
we need to add more assertions

Reply:
added!